lib/makara/connection_wrapper.rb

```diff
@@ -48,20 +57,55 @@ def _makara_custom_error_matchers
       @custom_error_matchers ||= (@config[:connection_error_matchers] || [])
     end
 
+    def _makara_connected?
+      _makara_connection.present?
+    rescue Makara::Errors::BlacklistConnection
+      false
+    end
+
+    def _makara_connection
+      current = @connection
+
+      if current
+        current
+      else # blacklisted connection or initial error
+        new_connection = @proxy.graceful_connection_for(@config)
+
+        # Already wrapped because of initial failure
+        if new_connection.is_a?(Makara::ConnectionWrapper)
+          _makara_blacklist!
+          raise Makara::Errors::BlacklistConnection.new(self, new_connection.initial_error)
+        else
+          @connection = new_connection
+          _makara_decorate_connection(new_connection)
+          new_connection
+        end
+      end
+    end
+
+    def execute(*args)
+      SQL_REPLACE.each do |find, replace|
+        if args[0] == find
+          args[0] = replace
+        end
+      end
+
+      _makara_connection.execute(*args)
+    end
+
     # we want to forward all private methods, since we could have kicked out from a private scenario
     def method_missing(m, *args, &block)
-      target = __getobj__
-      begin
-        target.respond_to?(m, true) ? target.__send__(m, *args, &block) : super(m, *args, &block)
-      ensure
-        $@.delete_if {|t| %r"\A#{Regexp.quote(__FILE__)}:#{__LINE__-2}:"o =~ t} if $@
+      if _makara_connection.respond_to?(m)
```

Comment:
It feels odd to me in this method to be using the `_makara_connection` method instead of the `@connection` variable. If i'm reading it right, you could get a different object back each time. think that's a problem?